tools/clang/lib/Sema/SemaHLSLDiagnoseTU.cpp

```diff
@@ -295,6 +301,25 @@ clang::FunctionDecl *ValidateNoRecursion(CallGraphWithRecurseGuard &callGraph,
   return nullptr;
 }
 
+class HLSLMethodCallDiagnoseVisitor
+    : public RecursiveASTVisitor<HLSLMethodCallDiagnoseVisitor> {
+public:
+  explicit HLSLMethodCallDiagnoseVisitor(Sema *S, const hlsl::ShaderModel *SM,
+                                         DXIL::ShaderKind EntrySK)
+      : sema(S), SM(SM), EntrySK(EntrySK) {}
+
+  bool VisitCXXMemberCallExpr(CXXMemberCallExpr *CE) {
+    sema->DiagnoseUsedHLSLMethodCall(CE->getMethodDecl(), CE->getExprLoc(), SM,
+                                     EntrySK);
+    return true;
+  }
+
+private:
+  clang::Sema *sema;
+  const hlsl::ShaderModel *SM;
+  DXIL::ShaderKind EntrySK;
```

Comment:
We should keep track of the root entry decl for use with a new note which identifies the entry that leads to the call of this method. This may also be used to check for other compatibility constraints, such as numthreads and broadcast launch constraints.  We could gather facts about the root entry once upon construction of this visitor class.

Reply:
Looks good, however, we still don't have a note pointing to an entry that was incompatible with the call, which would be useful when the call graph is deeper and there are multiple entries.